From b9819ae4a26caae8fd5e6e7658c8b68524ef31c1 Mon Sep 17 00:00:00 2001
From: Chris Lumens <clumens@redhat.com>
Date: Wed, 8 Apr 2015 14:59:42 +0800
Subject: [PATCH] Don't use the rpmdb to figure out upgrade target arch
 (#748119).

This will fail if /var is on a separate partition, since we do not have
the logic to mount additional partitions at this point.  Instead, just
run a chrooted /bin/arch which should always succeed.

Backport from https://github.com/rhinstaller/anaconda
715aceb15ddffd5f452c8ba9945b491a8a194f38

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 pyanaconda/storage/__init__.py | 25 +++++--------------------
 1 file changed, 5 insertions(+), 20 deletions(-)

diff --git a/pyanaconda/storage/__init__.py b/pyanaconda/storage/__init__.py
index 61da9b3..939c313 100644
--- a/pyanaconda/storage/__init__.py
+++ b/pyanaconda/storage/__init__.py
@@ -1480,28 +1480,13 @@ class Storage(object):
         return 0
 
 def getReleaseString(mountpoint):
-    relArch = None
     relName = None
     relVer = None
 
-    import rpm
-    iutil.resetRpmDb(mountpoint)
-    ts = rpm.TransactionSet(mountpoint)
-
-    # We get the arch from the initscripts package, but the version and name
-    # must come from reading the release file.
     try:
-        # pylint: disable-msg=E1101
-        mi = ts.dbMatch('provides', 'initscripts')
-    except Exception:
-        # This could happen in a variety of cases, but the biggest one is we're
-        # examining an installed system that doesn't use RPM.  Raise an
-        # exception for the caller to handle.
-        raise ValueError
-
-    for h in mi:
-        relArch = h['arch']
-        break
+        relArch = iutil.execWithCapture("arch", [], root=mountpoint).strip()
+    except:
+        relArch = None
 
     wrl_name = "Wind River Linux"
     filename_lsb = "%s/etc/lsb-release" % mountpoint
@@ -1577,8 +1562,8 @@ def findExistingRootDevices(anaconda, upgradeany=False):
                 device.teardown(recursive=True)
 
             if arch is None:
-                # we failed to determine the arch (for instance when there is a
-                # corrupted rpm database on the target system)
+                # we failed to determine the arch (for instance when we can't
+                # run a binary on the target system)
                 continue
 
             if upgradeany or \
-- 
1.9.1

